runtime/src/main/java/org/corfudb/runtime/view/stream/CompositeStreamAddressSpace.java

```diff
@@ -0,0 +1,338 @@
+                    }
+
+                    // Change lower bound if global address is not in the range of checkpoint addresses
+                    // i.e., optimize the space of addresses being synced.
+                    if (lowerBound < upperLimitAddressesCheckpointed && globalAddress > upperLimitAddressesCheckpointed) {
+                        lowerBound = upperLimitAddressesCheckpointed;
+                        pointerOnRegularStream = false;
+                    }
+                }
+            }
+
+            // Now, sync regular stream if requested tail is greater than the last synced address
+            // for this stream
+            if (regularSAS.getLastAddressSynced() < newTail) {
+                try {
+                    regularSAS.syncUpTo(globalAddress, newTail, lowerBound, readFn);
+                } catch (TrimmedException te) {
+                    // Throw exception if we were attempting to fulfill a snapshot transaction,
+                    // in the space of checkpoint addresses
+                    if (cpSAS.getLastAddressSynced() > globalAddress) {
+                        throw te;
+                    }
+                }
+
+                // If space of regular addresses is empty and CP stream contains addresses, point to CP.
+                if (regularSAS.isEmpty() && !cpSAS.isEmpty()) {
+                    pointerOnRegularStream = false;
+                }
+
+                // If we were able to build the regular log from the beginning with no trimmed
+                // exceptions, then remove the checkpoint entries, this way we will be able
+                // to provide snapshot transactions at the granularity of checkpoint(ed) entries
+                if (globalAddress < upperLimitAddressesCheckpointed &&
+                        regularSAS.getMin() > globalAddress) {
+                    pointerOnRegularStream = true;
+                    cpSAS.removeAddresses(cpSAS.getMax());
+                } else {
+                    regularSAS.setPointerToPosition(upperLimitAddressesCheckpointed);
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean containsAddress(long globalAddress) {
+        return regularSAS.containsAddress(globalAddress);
+    }
+
+    @Override
+    public long higher(long globalAddress, boolean forward) {
+        return regularSAS.higher(globalAddress, forward);
+    }
+
+    @Override
+    public long lower(long globalAddress, boolean forward) {
+        return regularSAS.lower(globalAddress, forward);
+    }
+
+    @Override
+    public int findAddresses(long oldTail, long newTail, Function<Long, ILogData> readFn) {
+        throw new UnsupportedOperationException("findAddresses");
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return this.regularSAS.isEmpty() && this.cpSAS.isEmpty();
+    }
+
+    @Override
+    public void setPointerToPosition(long address) {
+        this.regularSAS.setPointerToPosition(address);
```

Comment:
It looks like this method is used to advance the stream pointer, which is usually close to the tail, but calling indexOf on addresses will do a linear scan from the beginning of the array. Maybe you want to do it backwards instead if the array is large.